### PR TITLE
Relax UserFunctionVariable creation assert

### PR DESCRIPTION
**Pull request by voznesenskym**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #95667

Context: https://github.com/pytorch/pytorch/issues/95630#issuecomment-1447338984

cc @soumith @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

**Comment by ngimel** [comment]
Test?

File: torch/_dynamo/variables/functions.py [hunk 1]
```diff
@@ -113,7 +113,7 @@ def __init__(self, fn, is_constant=False, **kwargs):
             self.is_constant = False
 
         assert isinstance(
-            fn, (types.FunctionType, torch.jit.ScriptFunction)
+            fn, (types.MethodType, types.FunctionType, torch.jit.ScriptFunction)
```
Comment (mlazos) [inline torch/_dynamo/variables/functions.py:116]: Shouldn't this be handled by UserMethodVariable?

**Comment by jansel** [comment]
I'd be curious why this isn't being routed to `UserMethodVariable`.  It would be more clear with a test.

File: torch/_dynamo/variables/functions.py (hunk 1 above)
Comment (wconstab) [reply torch/_dynamo/variables/functions.py:116]: This is probably due to the issue i stumbled onto where promoting something to UMV opened up more inlining opportunities and introduced many graph breaks to models that currently didn't graph break.
https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/variables/builder.py#LL519C17-L519C17

IMO we should make a point of weeding out this kind of thing even if it means regressing (or disabling) some models and then hopefully fixing them back a better way.

**Comment by carmocca** [comment]
@jansel @voznesenskym https://github.com/pytorch/pytorch/issues/95630#issuecomment-1447015522 could serve as a test

**Review by ezyang — CHANGES_REQUESTED** [review]
The code clearly attempts to distinguish between UserMethodVariable and UserFunctionVariable. I want a clear explanation of how you pick between one class or the other.

**Comment by voznesenskym** [comment]
This isn't meant to land folks, it's to demonstrate the bug mentioned in https://github.com/pytorch/pytorch/issues/95630#issuecomment-1447338984

The root cause for *why* we get here like this is explained in that comment.

**Comment by ezyang** [comment]
> This isn't meant to land folks, it's to demonstrate the bug mentioned in https://github.com/pytorch/pytorch/issues/95630#issuecomment-1447338984

There's no way to tell from the PR desc that this was the intention

**Comment by voznesenskym** [comment]
> > This isn't meant to land folks, it's to demonstrate the bug mentioned in [#95630 (comment)](https://github.com/pytorch/pytorch/issues/95630#issuecomment-1447338984)
> 
> There's no way to tell from the PR desc that this was the intention

Yes, good feedback, the context bit could have been clearer.

**Comment by ngimel** [comment]
So how do we intend to fix #95630?

**Pull request closed without merging**